algopy_testing/src/algopy/primitives/bytes.py

```diff
@@ -0,0 +1,167 @@
+from __future__ import annotations
+
+import base64
+from collections.abc import Sequence
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from collections.abc import Iterator
+
+from algopy._constants import MAX_BYTES_SIZE
+from algopy.primitives.uint64 import UInt64
+
+# TypeError, ValueError are used for operations that are compile time errors
+# ArithmeticError and subclasses are used for operations that would fail during AVM execution
+
+
+class Bytes:
+    """
+    A python implementation of an AVM []byte
+    """
+
+    value: bytes  # underlying bytes value representing the []byte
+
+    def __init__(self, value: bytes = b"") -> None:
+        self.value = _as_bytes(value)
+
+    # @classmethod
+    # def from_int(cls, value: int = 0) -> Bytes:
+    #     return cls(value.to_bytes(8))
+
+    # @classmethod
+    # def from_str(cls, value: str) -> Bytes:
+    #     return cls(str.encode(value))
+
+    def __repr__(self) -> str:
+        return repr(bytes)
+
+    def __str__(self) -> str:
+        return str(self.value)
+
+    def __bool__(self) -> bool:
+        return bool(self.value)
+
+    def __add__(self, other: Bytes | bytes) -> Bytes:
+        """Concatenate Bytes with another Bytes or bytes literal
+        e.g. `Bytes(b"Hello ") + b"World"`."""
+        return (
```

Comment:
Despite the typing hints, I think we need to be conservative here and, attempt to convert other to bytes first, so the appropriate error can be thrown if the wrong type was passed. 

e.g.
```
result = self.value + as_bytes(other)
return _checked_result(result, "+")
```